sqlx-core/src/mysql/connection/stream.rs

```diff
@@ -137,14 +145,38 @@ impl MySqlStream {
         // https://dev.mysql.com/doc/dev/mysql-server/8.0.12/page_protocol_basic_packets.html
         // https://mariadb.com/kb/en/library/0-packet/#standard-packet
 
-        let mut header: Bytes = self.stream.read(4).await?;
-
-        let packet_size = header.get_uint_le(3) as usize;
-        let sequence_id = header.get_u8();
+        if let RecvPackageState::Done = self.recv_package_state {
+            self.recv_package_state = RecvPackageState::ReadingHeader {
+                data: [0; 4],
+                progress: 0,
+            };
+        }
 
-        self.sequence_id = sequence_id.wrapping_add(1);
+        if let RecvPackageState::ReadingHeader { data, progress } = &mut self.recv_package_state {
+            while *progress != 4 {
+                *progress += self.stream.read_some(&mut data[*progress..]).await?;
```

Comment:
Consider the case where the future is dropped or cancelled here.

Will the Buffstream (self.stream) then contain a payload without a header?

@mehcode proposes some code to avoid this in:

https://github.com/launchbadge/sqlx/issues/886#issuecomment-753688656

Reply:
Read_some will either read some data and return progress, or it will be dropped and have read data.
It it true that the buffstream will not contain the header, or only partially contain the header. Bu the read data is in the 
recv_package_state that is as much a part of the connection as the buff stream.

Reply:
As far ask I can tell, this makes a couple of assumptions.

1. sqlx will not drop the bufstream and later read from the stream.
2. there are never multiple threads trying to concurrently read from a given bufstream.

Is it certain these are both true?